### PR TITLE
git: worktree, support relative submodule URLs with proper path handling

### PR DESCRIPTION
**Pull request by cappyzawa**
### Core Implementation
- **Unified URL Resolution**: Implemented `resolveModuleURL()` using `internal/url` package for consistent URL handling
- **Protocol Support**: HTTPS, SSH (including SCP-like), and file:// protocols with relative URL resolution
- **Graceful Error Handling**: Missing origin remote doesn't cause errors, allowing flexible submodule configurations
- **Comprehensive Testing**: Added table-driven tests covering all supported protocols and edge cases

## Test Results

All WorktreeSuite tests pass successfully, including new relative URL resolution functionality:

```bash
$ go test -v -run "TestWorktreeSuite"
=== RUN   TestWorktreeSuite
=== RUN   TestWorktreeSuite/TestSubmodule
=== RUN   TestWorktreeSuite/TestSubmodules
=== RUN   TestWorktreeSuite/TestSubmodules_URLResolution
=== RUN   TestWorktreeSuite/TestSubmodules_URLResolution/HTTPS_origin_with_relative_URL
=== RUN   TestWorktreeSuite/TestSubmodules_URLResolution/SSH_origin_with_relative_URL
=== RUN   TestWorktreeSuite/TestSubmodules_URLResolution/HTTPS_origin_with_mixed_relative_and_absolute_URLs
=== RUN   TestWorktreeSuite/TestSubmodules_URLResolution/file_protocol_with_relative_URL
--- PASS: TestWorktreeSuite (10.67s)
PASS
ok  	github.com/go-git/go-git/v6	11.206s
```

Dedicated unit tests for the URL resolution function:

```bash
$ go test -v -run "TestResolveModuleURL"
=== RUN   TestResolveModuleURL
=== RUN   TestResolveModuleURL/https://github/user/repo2
=== RUN   TestResolveModuleURL/ssh://git@github.com:22/user/repo2.git
=== RUN   TestResolveModuleURL/git@github.com:user/repo2.git
=== RUN   TestResolveModuleURL/file://git/user/repo2.git
--- PASS: TestResolveModuleURL (0.00s)
PASS
ok  	github.com/go-git/go-git/v6	0.514s
```

**Related Issue**: https://github.com/go-git/go-git/issues/1380

### CODE REVIEW

**Comment by cappyzawa** [comment]
Before proceeding with additional changes to this PR, I'd like to ensure that all WorktreeSuite tests pass on the main branch first. 

I've created PR #1584 to fix the failing WorktreeSuite tests by adding required Author fields to CommitOptions. Once #1584 is merged, I'll rebase this branch and continue with any remaining work.

This approach ensures a clean review environment without unrelated test failures.

File: worktree_test.go [hunk 1]
```diff
@@ -1585,6 +1593,109 @@ func (s *WorktreeSuite) TestSubmodules() {
...
+
+	err = util.WriteFile(fs, ".gitmodules", []byte(gitmodulesContent), 0644)
+	s.NoError(err)
+
+	_, err = w.Submodules()
+	s.Error(err)
+	s.Contains(err.Error(), "failed to resolve relative submodule URL")
+	s.Contains(err.Error(), "file:// protocol requires .git suffix")
+}
+
```
Comment (pjbgf) [inline worktree_test.go:1758]: Add a table driven test focusing on the func that resolves URL, so that it is easier to catch lack of coverage/use cases. Here's a suggestion:

```suggestion
func TestResolveModuleURL(t *testing.T) {
	tests := []struct {
		originURL string
		moduleURL string
		want      string
		wantErr   error
	}{
		{
			originURL: "https://github/user/repo1",
			moduleURL: "../repo2",
			want:      "https://github/user/repo2",
		},
		{
			originURL: "https://github/user/repo1",
			moduleURL: "https://github/user/repo3",
			want:      "https://github/user/repo3",
		},
		{
			originURL: "ssh://git@github.com:22/user/repo1.git",
			moduleURL: "../repo2.git",
			want:      "ssh://git@github.com:22/user/repo2.git",
		},
		{
			originURL: "ssh://git@github.com:22/user/repo1.git",
			moduleURL: "ssh://git@github.com:22/user/repo3.git",
			want:      "ssh://git@github.com:22/user/repo3.git",
		},
		{
			originURL: "git@github.com:2222:user/repo1.git",
			moduleURL: "../repo2.git",
			want:      "git@github.com:2222:user/repo2.git",
		},
		{
			originURL: "git@github.com:user/repo1.git",
			moduleURL: "../repo2.git",
			want:      "git@github.com:user/repo2.git",
		},
		{
			originURL: "git@github.com:user/repo1.git",
			moduleURL: "git@github.com:user/repo3.git",
			want:      "git@github.com:user/repo3.git",
		},
		{
			originURL: "file://git/user/repo1.git",
			moduleURL: "../repo2.git",
			want:      "file://git/user/repo2.git",
		},
		{
			originURL: "file://git/user/repo1.git",
			moduleURL: "file://git/user/repo3.git",
			want:      "file://git/user/repo3.git",
		},
		{
			originURL: "/git/user/repo1.git",
			moduleURL: "../repo2.git",
			want:      "/git/user/repo2.git",
		},
		{
			originURL: "/git/user/repo1.git",
			moduleURL: "/git/user/repo3.git",
			want:      "/git/user/repo3.git",
		},
		{
			originURL: "",
			moduleURL: "../repo2/.git",
			want:      "../repo2/.git",
		},
	}

	for _, tc := range tests {
		t.Run(tc.want, func(t *testing.T) {
			got, err := resolveModuleURL(tc.originURL, tc.moduleURL)
			if tc.wantErr == nil {
				assert.NoError(t, err)
				assert.Equal(t, tc.want, got)
			} else {
				assert.Empty(t, got)
				assert.ErrorIs(t, err, tc.wantErr)
			}
		})
	}
}
```

File: worktree.go [hunk 2]
```diff
@@ -897,6 +900,130 @@ func (w *Worktree) Submodule(name string) (*Submodule, error) {
...
+	if originRemote == nil {
+		return "", fmt.Errorf("no origin remote found")
+	}
+
+	if len(originRemote.c.URLs) == 0 {
+		return "", fmt.Errorf("origin remote has no URLs configured")
+	}
+
+	return originRemote.c.URLs[0], nil
+}
```
Comment (pjbgf) [inline worktree.go]: Let's inline this instead - please refer to suggestions on `Submodules()`.
```suggestion
```

File: worktree.go [hunk 3]
```diff
@@ -897,6 +900,130 @@ func (w *Worktree) Submodule(name string) (*Submodule, error) {
...
+	// For file:// URLs like file:///path/to/repo.git
+	if !strings.HasSuffix(relativeURL, ".git") {
+		return "", fmt.Errorf("file:// protocol requires .git suffix, got: %s", relativeURL)
+	}
+
+	baseDir := path.Dir(baseEndpoint.Path)
+	resolvedPath := path.Join(baseDir, relativeURL)
+
+	return fmt.Sprintf("file://%s", resolvedPath), nil
+}
```
Comment (pjbgf) [inline worktree.go]: Let's have a single source of truth and lean in on Go's std lib as much as possible, instead of creating a func per type of URL. 
```suggestion
```

File: worktree.go [hunk 4]
```diff
@@ -897,6 +900,130 @@ func (w *Worktree) Submodule(name string) (*Submodule, error) {
...
+		return w.resolveHTTPRelativeURL(originURL, relativeURL)
+	case "ssh":
+		return w.resolveSSHRelativeURL(baseEndpoint, relativeURL, originURL)
+	case "file":
+		return w.resolveFileRelativeURL(baseEndpoint, relativeURL)
+	default:
+		// For other protocols, we don't support relative URL resolution
+		return "", fmt.Errorf("relative URL resolution not supported for protocol: %s", baseEndpoint.Protocol)
+	}
+}
```
Comment (pjbgf) [inline worktree.go]: What about:

```suggestion
// resolveModuleURL resolves relative URLs based on the originURL.
// The moduleURL is returned verbatim when originURL is empty or moduleURL is an absolute URL.
func resolveModuleURL(originURL, moduleURL string) (string, error) {
	if originURL == "" {
		return moduleURL, nil
	}

	if !strings.HasPrefix(moduleURL, "../") && !strings.HasPrefix(moduleURL, "./") {
		return moduleURL, nil
	}

	if !giturl.MatchesScheme(originURL) && giturl.MatchesScpLike(originURL) {
		user, host, portStr, path := giturl.FindScpLikeComponents(originURL)

		path = filepath.Join(path, moduleURL)
		if portStr != "" {
			portStr += ":"
		}

		return fmt.Sprintf("%s@%s:%s%s", user, host, portStr, path), nil
	}

	base, err := url.Parse(originURL)
	if err != nil {
		return "", err
	}

	base.Path = filepath.Join(base.Path, moduleURL)

	return base.String(), nil
}
```

The only use case that needs to be handled differently is SCP-like URLs, everything else seems to work fine with `url.Parse`.

File: worktree.go [hunk 5]
```diff
@@ -910,8 +1037,43 @@ func (w *Worktree) Submodules() (Submodules, error) {
...
+	var originURL string
+	for _, s := range m.Submodules {
+		if s.URL != "" && isRelativeURL(s.URL) {
+			originURL, err = w.getOriginRemoteURL()
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+	}
```
Comment (pjbgf) [inline worktree.go]: We could try to get the default remote, if not found we skip the resolution (by leaving the URL empty):

```suggestion
	var originURL string
	if origin, err := w.r.Remote(DefaultRemoteName); err == nil {
		if origin.c != nil && len(origin.c.URLs) > 0 {
			originURL = origin.c.URLs[0]
		}
	}
```

File: worktree.go [hunk 6]
```diff
@@ -910,8 +1037,43 @@ func (w *Worktree) Submodules() (Submodules, error) {
...
+		resolvedURL, err := w.resolveRelativeSubmoduleURL(s.URL, originURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve relative submodule URL %q: %w", s.URL, err)
+		}
+		copiedSubmodule.URL = resolvedURL
+		if copiedConfig != nil {
+			copiedConfig.URL = resolvedURL
+		}
+
+		l = append(l, w.newSubmodule(&copiedSubmodule, copiedConfig))
```
Comment (pjbgf) [inline worktree.go]: Making `resolveModuleURL` the source of truth, means that we don't need to have any logic here around whether or not the URL is relative:

```suggestion
		sub := w.newSubmodule(s, c.Submodules[s.Name])
		cfg := sub.Config()

		resolvedURL, err := resolveModuleURL(originURL, cfg.URL)
		if err != nil {
			return nil, fmt.Errorf("failed to resolve submodule URL %q: %w", s.URL, err)
		}

		cfg.URL = resolvedURL
		l = append(l, sub)
```

File: worktree_test.go [hunk 7]
```diff
@@ -1585,6 +1593,109 @@ func (s *WorktreeSuite) TestSubmodules() {
...
+	gitmodulesContent := `[submodule "relative"]
+	path = relative
+	url = ../relative-repo.git`
+
+	err = util.WriteFile(fs, ".gitmodules", []byte(gitmodulesContent), 0644)
+	s.NoError(err)
+
+	_, err = w.Submodules()
+	s.Error(err)
+	s.Contains(err.Error(), "no origin remote found")
```
Comment (pjbgf) [inline worktree_test.go]: We don't want to error if origin does not exist.

```suggestion
	s.NoError(err)
```

File: worktree_test.go [hunk 8]
```diff
@@ -1585,6 +1593,109 @@ func (s *WorktreeSuite) TestSubmodules() {
...
+	path = relative
+	url = ../relative-repo`
+
+	err = util.WriteFile(fs, ".gitmodules", []byte(gitmodulesContent), 0644)
+	s.NoError(err)
+
+	_, err = w.Submodules()
+	s.Error(err)
+	s.Contains(err.Error(), "failed to resolve relative submodule URL")
+	s.Contains(err.Error(), "file:// protocol requires .git suffix")
```
Comment (pjbgf) [inline worktree_test.go]: Although we might not be able to support the loading of this depending on the fs being used, for the purposes of the submodule logic, this does not need to error:

```suggestion
	s.NoError(err)
```

**Review by pjbgf — CHANGES_REQUESTED** [review]
@cappyzawa thank you for the quick turn around. Below are a few suggestions.

File: worktree_test.go (hunk 1 above)
Comment (cappyzawa) [reply worktree_test.go:1758]: Thanks for your suggestion 👍 

Fixed via 6d6339a3 - Added comprehensive table-driven test TestResolveModuleURL covering all the suggested test cases including edge cases like empty origin URL and non-relative URLs. This provides better coverage and makes it easier to add new test cases.

File: worktree.go (hunk 2 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via 6d6339a3 .

File: worktree.go (hunk 3 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via 6d6339a3 and 7433bf0 .

File: worktree.go (hunk 4 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via 6d6339a3 - Thank you for the excellent suggestion.

File: worktree.go (hunk 5 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via 6d6339a3 .

File: worktree.go (hunk 6 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via 6d6339a3 - Thank you. It became much more concise.

File: worktree_test.go (hunk 8 above)
Comment (cappyzawa) [reply worktree_test.go]: Fixed via 6d6339a3 - These edge cases are now covered by the comprehensive TestResolveModuleURL table-driven test, so these separate test functions were removed to avoid duplication.

File: worktree_test.go (hunk 7 above)
Comment (cappyzawa) [reply worktree_test.go]: Fixed via 6d6339a3 - These edge cases are now covered by the comprehensive TestResolveModuleURL table-driven test, so these separate test functions were removed to avoid duplication.

**Comment by cappyzawa** [comment]
@pjbgf Thank you for the detailed review feedback! All review comments have been addressed with the specific commit hashes provided in the individual comment responses.

## Status Update

Initially, I was going to prioritize PR #1584 (test improvements) to resolve test failures in my local environment first. However, I discovered that I can work around the environment issues by adjusting my local git configuration, so I've decided to move forward with this PR instead.

Ready for re-review! I've removed the draft status to proceed with the review process.

**Comment by cappyzawa** [comment]
**Note on commit history**: After approval, I will clean up the commit history by squashing fixup commits and removing the revert commit to create a clean merge-ready state.

File: worktree.go [hunk 9]
```diff
@@ -897,6 +899,31 @@ func (w *Worktree) Submodule(name string) (*Submodule, error) {
...
+		if portStr != "" {
+			portStr += ":"
+		}
+		return fmt.Sprintf("%s@%s:%s%s", user, host, portStr, path), nil
+	}
+	base, err := url.Parse(originURL)
+	if err != nil {
+		return "", err
+	}
+	base.Path = filepath.Join(base.Path, moduleURL)
```
Comment (pjbgf) [inline worktree.go]: This might fix the failures in Windows:
```suggestion
	base.Path = path.Join(base.Path, moduleURL)
```

File: worktree.go [hunk 10]
```diff
@@ -897,6 +899,31 @@ func (w *Worktree) Submodule(name string) (*Submodule, error) {
...
+	if !strings.HasPrefix(moduleURL, "../") && !strings.HasPrefix(moduleURL, "./") {
+		return moduleURL, nil
+	}
+	if !giturl.MatchesScheme(originURL) && giturl.MatchesScpLike(originURL) {
+		user, host, portStr, path := giturl.FindScpLikeComponents(originURL)
+		path = filepath.Join(path, moduleURL)
+		if portStr != "" {
+			portStr += ":"
+		}
+		return fmt.Sprintf("%s@%s:%s%s", user, host, portStr, path), nil
```
Comment (pjbgf) [inline worktree.go]: For SCP-like paths, this might not be enough, but it should be easier to debug this without having to run in Windows:

```suggestion
		user, host, portStr, p := giturl.FindScpLikeComponents(originURL)
		p = path.Join(path, moduleURL)
		if portStr != "" {
			portStr += ":"
		}
		return fmt.Sprintf("%s@%s:%s%s", user, host, portStr, p), nil
```

File: worktree.go (hunk 9 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via [2fbaf8f](https://github.com/go-git/go-git/pull/1581/commits/2fbaf8fabb8d57bff1884ef6958dacec8207e4d2) .

File: worktree.go (hunk 10 above)
Comment (cappyzawa) [reply worktree.go]: Fixed via [2fbaf8f](https://github.com/go-git/go-git/pull/1581/commits/2fbaf8fabb8d57bff1884ef6958dacec8207e4d2) .

**Comment by cappyzawa** [comment]
Hi @pjbgf,

Thank you for the thorough review! I've completed the commit log cleanup and squashed everything into a single commit.

I have to admit, you essentially implemented this entire feature for me through your detailed suggestions and concrete code examples. Your approach using the internal/url package was far superior to my initial complex implementation attempt. 

Thank you for taking the time to provide such excellent guidance and for essentially walking me through the proper solution. This was an incredible learning experience!

The PR should now be ready for merge.

**Review by pjbgf — APPROVED** [review]
@cappyzawa thank you very much for working on this. :bow:
